system/HTTP/URL.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+/**
+ * This file is part of the CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\HTTP;
+
+use CodeIgniter\HTTP\IncomingRequest;
+use CodeIgniter\Router\Exceptions\RouterException;
+use Config\App;
+use Config\Services;
+use InvalidArgumentException;
+
+/**
+ * URI wrapper to work with complete project URLs.
+ * This class creates immutable instances.
+ */
+final class URL
+{
+	/**
+	 * Explicit URL to use for current.
+	 *
+	 * @var static|null
+	 */
+	private static $current;
+
+	/**
+	 * Underlying URI instance
+	 *
+	 * @var URI
+	 */
+	private $uri;
+
+	/**
+	 * Path relative to baseURL
+	 *
+	 * @var string
+	 */
+	private $relativePath;
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Creates the base URL.
+	 *
+	 * @return static
+	 */
+	public static function base()
+	{
+		return static::public('');
```

Comment:
Since the class is `final`, then inheritance is no more possible. So, there's no need to use the `static` keyword. Instead use `self`.
```suggestion
		return self::public('');
```

Reply:
On the topic, how do you think the class should be scoped? [Originally](https://github.com/codeigniter4/CodeIgniter4/pull/4647/commits/42a60ce50059ec4f492fd72078b6b3a972853ee2) I had a normal class with `final` static and constructor methods (hence the leftover `static` references here) but I decided that nobody would go through the hassle of extending for the few allowed methods so made the whole thing `final`.